src/vm/jitinterface.cpp

```diff
@@ -4585,6 +4585,76 @@ BOOL CEEInfo::areTypesEquivalent(
 }
 
 /*********************************************************************/
+// See if a cast from fromClass to toClass will succeed, fail, or needs
+// to be resolved at runtime.
+TypeCompareState CEEInfo::compareTypesForCast(
+        CORINFO_CLASS_HANDLE        fromClass,
+        CORINFO_CLASS_HANDLE        toClass)
+{
+    // Stub for now
+    return TypeCompareState::May;
+}
+
+/*********************************************************************/
+// See if types represented by cls1 and cls2 compare equal, not
+// equal, or the comparison needs to be resolved at runtime.
+TypeCompareState CEEInfo::compareTypesForEquality(
+        CORINFO_CLASS_HANDLE        cls1,
+        CORINFO_CLASS_HANDLE        cls2)
+{
+    CONTRACTL {
+        SO_TOLERANT;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    TypeCompareState result = TypeCompareState::May;
+
+    JIT_TO_EE_TRANSITION();
+
+    TypeHandle hnd1 = (TypeHandle) cls1;
+    TypeHandle hnd2 = (TypeHandle) cls2;
+
+    // If neither type is a canonical subtype, type handle comparision suffices
+    if (!hnd1.IsCanonicalSubtype() && !hnd2.IsCanonicalSubtype())
+    {
+        result = (hnd1 == hnd2? TypeCompareState::Must : TypeCompareState::MustNot);
+    }
+    // If either or both types are canonical subtypes, we can sometimes prove inequality.
+    else
+    {
+        // If either is a value type then the types cannot
+        // be equal unless the type defs are the same
+        if (hnd1.IsValueType() || hnd2.IsValueType())
+        {
+            if (!hnd1.AsMethodTable()->HasSameTypeDefAs(hnd2.AsMethodTable()))
```

Comment:
There are "native value types" for which IsValueType() returns true, but they are not backed by MethodTable. You may want to check for !IsTypeDesc() before calling AsMethodTable() here.

Reply:
How about:
```C++
        if ((hnd1.IsValueType() && !hnd1.IsNativeValueType())
            || (hnd2.IsValueType() && !hnd2.IsNativeValueType()))
```

Reply:
This can be still true if one of them native value. It may be best to handle this by calling `GetMethodTable` instead of `AsMethodTable`. Something like:

```
      if (hnd1.IsValueType() || hnd2.IsValueType())
      {
          if (!hnd1.GetMethodTable()->HasSameTypeDefAs(hnd2.GetMethodTable()))
```

Reply:
So there are classes that aren't value types but are native value types? Interesting. Is there some simple way to create these kinds of classes?

Any extra complications on desktop? I need to implement at least part of this there...

Reply:
The situation that I was worried about does not actually need to involve native value types. For example: hnd1 is value type and hnd2 is array. `if (hnd1.IsValueType() || hnd2.IsValueType())` will be true, and we would end up calling AsMethodTable on array TypeDesc.

> So there are classes that aren't value types but are native value types? 

All native value types are value types too.